### ISSUE TITLE
Meta-variable solutions contain subterms with the wrong modality

### ISSUE THREAD

**Issue by nad** [agda/agda#4727]
This problem was discovered by @Saizan.

The following code is accepted:
```agda
{-# OPTIONS --without-K --safe #-}

open import Agda.Builtin.Equality
open import Agda.Builtin.Sigma

private
  variable
    A B   : Set
    P Q   : A → Set
    f x y : A

id : A → A
id x = x

_∘_ : {B : A → Set} {C : {x : A} → B x → Set} →
      (∀ {x} (y : B x) → C y) → (g : (x : A) → B x) →
      ((x : A) → C (g x))
f ∘ g = λ x → f (g x)

Σ-map : (f : A → B) → (∀ {x} → P x → Q (f x)) →
        Σ A P → Σ B Q
Σ-map f g (x , y) = f x , g y

record Erased (@0 A : Set) : Set where
  constructor [_]
  field
    @0 erased : A

open Erased

cong : (f : A → B) → x ≡ y → f x ≡ f y
cong f refl = refl

Contractible : Set → Set
Contractible A = Σ A λ x → ∀ y → x ≡ y

Contractibleᴱ : Set → Set
Contractibleᴱ A = Σ A λ x → Erased (∀ y → x ≡ y)

_⁻¹_ : (A → B) → B → Set
f ⁻¹ y = Σ _ λ x → f x ≡ y

_⁻¹ᴱ_ : {@0 B : Set} → @0 (A → B) → @0 B → Set
f ⁻¹ᴱ y = Σ _ λ x → Erased (f x ≡ y)

Is-equivalence : (A → B) → Set
Is-equivalence f = ∀ y → Contractible (f ⁻¹ y)

Is-equivalenceᴱ : @0 (A → B) → Set
Is-equivalenceᴱ f = ∀ y → Contractibleᴱ (f ⁻¹ᴱ y)

@0 Contractibleᴱ→Contractible : Contractibleᴱ A → Contractible A
Contractibleᴱ→Contractible = Σ-map id erased

⁻¹→⁻¹ᴱ : f ⁻¹ y → f ⁻¹ᴱ y
⁻¹→⁻¹ᴱ = Σ-map id (λ x → [ x ])

@0 ⁻¹ᴱ→⁻¹ : f ⁻¹ᴱ y → f ⁻¹ y
⁻¹ᴱ→⁻¹ = Σ-map id erased

@0 Is-equivalenceᴱ→Is-equivalence :
  Is-equivalenceᴱ f → Is-equivalence f
Is-equivalenceᴱ→Is-equivalence eq y =
    ⁻¹ᴱ→⁻¹ (fst (eq y))
  , (cong ⁻¹ᴱ→⁻¹ ∘ (erased (snd (eq y)) ∘ ⁻¹→⁻¹ᴱ))

record _≃_ (A B : Set) : Set where
  constructor ⟨_,_⟩
  field
    to             : A → B
    is-equivalence : Is-equivalence to

  from : B → A
  from y = fst (fst (is-equivalence y))

  right-inverse-of : ∀ y → to (from y) ≡ y
  right-inverse-of y = snd (fst (is-equivalence y))

  irrelevance : ∀ y (p : to ⁻¹ y) → (from y , right-inverse-of y) ≡ p
  irrelevance y = snd (is-equivalence y)

record _≃ᴱ_ (A B : Set) : Set where
  constructor ⟨_,_⟩
  field
    to             : A → B
    is-equivalence : Is-equivalenceᴱ to

  from : B → A
  from y = fst (fst (is-equivalence y))

  @0 equivalence : A ≃ B
  equivalence =
    ⟨ to , Is-equivalenceᴱ→Is-equivalence is-equivalence ⟩

record Erased-proofs {A B : Set} (to : A → B) (from : B → A) : Set where
  field
    right-inverse-of : ∀ y → to (from y) ≡ y
    irrelevance      : ∀ y (p : to ⁻¹ y) →
                       (from y , right-inverse-of y) ≡ p

[proofs] :
  (A≃B : A ≃ B) → Erased-proofs (_≃_.to A≃B) (_≃_.from A≃B)
[proofs] A≃B .Erased-proofs.right-inverse-of =
  _≃_.right-inverse-of A≃B
[proofs] A≃B .Erased-proofs.irrelevance =
  _≃_.irrelevance A≃B

[≃]→≃ᴱ :
  {to : A → B} {from : B → A} →
  @0 Erased-proofs to from →
  A ≃ᴱ B
[≃]→≃ᴱ {to = to} {from = from} ep =
  ⟨ to
  , (λ y → (from y , [ Erased-proofs.right-inverse-of ep y ])
         , [ cong ⁻¹→⁻¹ᴱ ∘ (Erased-proofs.irrelevance ep y ∘ ⁻¹ᴱ→⁻¹) ])
  ⟩

Σ-cong :
  (A≃B : A ≃ B) →
  (∀ x → P x ≃ Q (_≃_.to A≃B x)) →
  Σ A P ≃ Σ B Q
Σ-cong = {!!}

Σ-congᴱ :
  (A≃B : A ≃ B) →
  (∀ x → P x ≃ᴱ Q (_≃_.to A≃B x)) →
  Σ A P ≃ᴱ Σ B Q
Σ-congᴱ f g =
  [≃]→≃ᴱ {to = {!!}} {from = {!!}}
    ([proofs] (Σ-cong f (_≃ᴱ_.equivalence ∘ g)))
```
The last two meta-variables are solved. However, the solutions include `_≃ᴱ_.equivalence`, which is erased:
```
?1 := _≃_.to (Σ-cong f (λ x → _≃ᴱ_.equivalence (g x)))
?2 := _≃_.from (Σ-cong f (λ x → _≃ᴱ_.equivalence (g x)))
```

**Comment by Saizan** [comment]
I have reduced the test case a bit.
```agda
private
  variable
    A B   : Set
    P Q   : A → Set

postulate
  Σ : (A : Set) (B : A → Set) -> Set
  _∘_ : {A : Set} {B : A → Set} {C : {x : A} → B x → Set} →
      (∀ {x} (y : B x) → C y) → (g : (x : A) → B x) →
      ((x : A) → C (g x))

postulate
  _≃_ : (A B : Set) → Set

module _ (e : A ≃ B) where
 postulate
  to             : A → B
  from : B → A

postulate
  _≃ᴱ_ : (A B : Set) → Set

  @0 equivalence : A ≃ᴱ B → A ≃ B

  Erased-proofs : {A B : Set} (to : A → B) (from : B → A) → Set
  [proofs] : (A≃B : A ≃ B) → Erased-proofs (to A≃B) (from A≃B)
  [≃]→≃ᴱ :
    {to : A → B} {from : B → A} →
    @0 Erased-proofs to from →
    A ≃ᴱ B
  Σ-cong :
    (A≃B : A ≃ B) →
    (∀ x → P x ≃ Q (to A≃B x)) →
    Σ A P ≃ Σ B Q

Σ-congᴱ : ∀ {A B P Q} →
  (A≃B : A ≃ B) →
  (∀ x → P x ≃ᴱ Q (to A≃B x)) →
  Σ A P ≃ᴱ Σ B Q
Σ-congᴱ {A} {B} {P} {Q} f g =
  [≃]→≃ᴱ {to = _} {from = _}
    (([proofs] ((Σ-cong {P = P} f (_∘_ equivalence g)))))
```

The most puzzling part is that if we give `{Q = Q}` to `Σ-cong` then the underscores are not solved.

**Comment by nad** [comment]
I tweaked the test case a little:
```agda
postulate
  comp :
    (A : Set) (B C : A → Set) →
    ((x : A) → B x → C x) → ((x : A) → B x) → (x : A) → C x

  _≃_  : Set → Set → Set
  to   : (A B : Set) → A ≃ B → A → B
  from : (A B : Set) → A ≃ B → B → A

  Σ      : (A : Set) → (A → Set) → Set
  Σ-cong :
    (A B : Set) (P : A → Set) (Q : B → Set) (A≃B : A ≃ B) →
    (∀ x → P x ≃ Q (to A B A≃B x)) → Σ A P ≃ Σ B Q

  Proofs : (A B : Set) → (A → B) → (B → A) → Set
  proofs : (A B : Set) (A≃B : A ≃ B) →
           Proofs A B (to A B A≃B) (from A B A≃B)
  [≃]→≃  : (A B : Set) (to : A → B) (from : B → A) →
           @0 Proofs A B to from → A ≃ B

  @0 erased : (A B : Set) → A ≃ B → A ≃ B

bad :
  (A B : Set) (P : A → Set) (Q : B → Set) (A≃B : A ≃ B) →
  (∀ x → P x ≃ Q (to A B A≃B x)) →
  Σ A P ≃ Σ B Q
bad A B P Q f g =
  [≃]→≃ (Σ A P) (Σ B Q) _ _
    (proofs (Σ A P) (Σ B _)
       (Σ-cong A B P _ f
          (comp
             A
             (λ x → P x ≃ Q (to A B f x))
             (λ x → P x ≃ Q (to A B f x))
             (λ x → erased (P x) (Q (to A B f x)))
             g)))
```
I looked at Agda's debug output. The meta-variable corresponding to the first underscore in `bad` is instantiated to a term that contains the meta-variable `_49`. Later `_49` is instantiated to a term that contains `erased` (there is a boxed constraint between `_49` and the subterm starting with `comp`).

How does Agda keep track of whether a meta-variable's solution can contain erased terms?

**Comment by nad** [comment]
@jespercockx, the problem is still there if `--double-check` is used. Doesn't the double-checker check things after meta-variables have been instantiated?

**Comment by jespercockx** [comment]
Uh, it seems `checkInternal` only checks hiding & relevance of arguments, not other modalities

https://github.com/agda/agda/blob/08191e66e65a4870bf43e21007909fc03150eabf/src/full/Agda/TypeChecking/CheckInternal.hs#L276-L285

I have the feeling this *might* explain why `--double-check` doesn't catch anything here ;)

**Comment by jespercockx** [comment]
I've fixed this issue with `checkInternal` (see https://github.com/jespercockx/agda/tree/issue4727) but that doesn't seem to change anything for the test case here. 

Now that I think about it, it makes sense, since the issue here is more of a global violation of the erasure rules rather than a local one: probably the internal has the right annotation, but this annotation is lying about the actual contents of the argument.

**Comment by jespercockx** [comment]
The check that the solution of a non-erased metavariable should not contain any erased subterms, is done by the Occurs checker (Agda.TypeChecking.MetaVars.Occurs). However, while there is a check for erased definitions and erased (regular) variables, the check for erased metavariables seems to be missing. The strange thing is that I cannot find the corresponding check for irrelevance either, so I might still be looking in the wrong place.

**Comment by jespercockx** [comment]
The test case still passes when replacing erasure by irrelevance, so perhaps the problem is not just there for erasure.

**Comment by jespercockx** [comment]
Bingo, adding this check to the occurs checker makes the file fail with unsolved metas (both with erasure and irrelevance). Now we'll just have to see how many things this breaks in other test cases...

**Comment by nad** [comment]
> Now we'll just have to see how many things this breaks in other test cases...

Did anything break?

**Comment by jespercockx** [comment]
Yes, just postponing when there's a meta doesn't quite work. It might be better to instead 'promote' the meta from erased to non-erased (similar to pruning) but it'll take a bit more time to actually implement that.

**Comment by Saizan** [comment]
@jespercockx have you pushed your attempt somewhere?

**Comment by jespercockx** [comment]
https://github.com/jespercockx/agda/tree/issue4727

EDIT: wait no that's not the right one

EDIT 2: now it's actually there

**Comment by nad** [comment]
@Saizan, did you look at this?

**Comment by Saizan** [comment]
I fixed the unsolved metas in `make succeed` by 1ec9948 . 
We still get an IMPOSSIBLE on `Issue3390.agda` though:

```
An internal error has occurred. Please report this as a bug.
Location of the error: __IMPOSSIBLE__, called at src/full/Agda/TypeChecking/Monad/Context.hs:130:32 in Agda-2.6.2-LvRmYY9su5wDW05FKoAKoV:Agda.TypeChecking.Monad.Context
```

Something to do with checkpoints:
```
-- | Get the substitution from the context at a given checkpoint to the current context.                                            
checkpointSubstitution :: MonadTCEnv tcm => CheckpointId -> tcm Substitution
checkpointSubstitution = maybe __IMPOSSIBLE__ return <=< checkpointSubstitution'
```

@jespercockx got some idea of what could be going wrong?

**Comment by jespercockx** [comment]
It seems it's somehow triggered by the generalizable variables; if I replace them with postulates or quantify over `q` explicitly in the type of `d` then the error disappears. I tried to exclude generalizable metas from the promotion process, however the problem persists. I'll have to take a deeper look.

**Comment by nad** [comment]
@UlfNorell recently fixed a problem related to modalities and generalisable variables (#5058). Perhaps it makes sense to rebase your branch.

**Comment by jespercockx** [comment]
I just did a rebase but unfortunately that didn't make the problem disappear :(

**Comment by nad** [comment]
This has not been completely fixed yet. The following code is accepted:
```agda
data D : Set where
  @0 c : D

postulate
  P : D → Set
  p : (x : D) → P x

bad : P c
bad = p _
```
The result of normalising `bad` is `p c`.

The following variant of the code is rejected:
```agda
postulate
  D    : Set
  @0 c : D
  P    : D → Set
  p    : (x : D) → P x

bad : P c
bad = p _
```
```
Cannot instantiate the metavariable _5 to solution c
since (part of) the solution was created in an erased context
when checking that the expression p _ has type P c
```

**Comment by jespercockx** [comment]
The fix I implemented only prevents the case where a non-erased metavariable *indirectly* gets an erased solution via another (erased) metavariable. It seems that when erased constructors were implemented, the *direct* check for instantiating a metavariable with a solution containing an erased constructor was not implemented. The check should be inserted at https://github.com/agda/agda/blob/75d82f728dbb65a826e09a37dc313a7a132e1933/src/full/Agda/TypeChecking/MetaVars/Occurs.hs#L471, there should be a call to `definitionCheck` like in the case for `def` just above it.
